hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -387,22 +387,27 @@ func (r *HostedClusterReconciler) reconcile(ctx context.Context, req ctrl.Reques
 	}
 
 	reportLimitedSuportEnabled(hcluster)
-	reportSilecedAlerts(hcluster)
+	reportSilencedAlerts(hcluster)
 	reportProxyConfig(hcluster)
 
 	// If deleted, clean up and return early.
 	if !hcluster.DeletionTimestamp.IsZero() {
 		reportHostedClusterGuestCloudResourcesDeletionDuration(hcluster)
+		reportHostedClusterDeletionDuration(hcluster, r.Clock)
 
 		// Keep trying to delete until we know it's safe to finalize.
 		completed, err := r.delete(ctx, hcluster)
 		if err != nil {
 			return ctrl.Result{}, fmt.Errorf("failed to delete hostedcluster: %w", err)
 		}
+
 		if !completed {
 			log.Info("hostedcluster is still deleting", "name", req.NamespacedName)
 			return ctrl.Result{RequeueAfter: clusterDeletionRequeueDuration}, nil
 		}
+
+		defer hcmetrics.StopReporting(hcluster)
```

Comment:
if `controllerutil.RemoveFinalizer` this will delete metrics before the hostedCluster actual deletion duration has been reported here:
https://github.com/openshift/hypershift/blob/b98cb4a246e8bdabfbc7154cae20b0f16cda8260/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#LL414C3-L414C38

Reply:
Indeed we should probably not delete the time series yet if we can successfully remove the finalizer.
Remark that the impact is quite limited, here is what happens:
1. HC finalizer is removed:
2. Times series for the HC are all ended
3. HC is reconciled (because of the finalizer removal)
4. Time series for `hypershift_cluster_upgrade_duration_seconds` metric (and probably some other metrics) get  opened again and populated.
5. Times series for the HC are all ended
Nevertheless you are right, we should change that; thanks for reporting

Reply:
I forgot to say but I encountered the "failed to remove finalizer from hostedcluster" while testing.
It seems that, even after the finalizer is removed, it may take a bit of time to definitely remove the `hostedcluster` object.
During that time the `reconcile` method may be entered again:
1. At the beginning the `hostedcluster` object may still exist in database; hence `hcluster` variable get correctly loaded and set.
2. Just before the deletion block, we will call all the "report" methods which are defining again the time series we just deleted.
3. The `r.delete` call will do nothing as all the objects we need to delete prior the `hostedcluster` are already deleted.
4. Then trying to remove the finalizer again will fail as the `hostedcluster` has finally been removed from DB.

As removing the finalizer seems to only fail when the `hostedcluster` no longer exists in DB (the `hypershift-operator` would have encountered other update issues before going that far in the delete process if it wasn't able to edit the `hostedcluster` object in DB), it is ok to stop defining the metrics in that case.

Reply:
@muraee, I actually misread your above message.
I need to stop the time series when we successfully delete the finalizer or when we do not delete it successfully.. in order to accommodate the weird case / race condition explained above.

Reply:
So I added back the `defer` expression.